oak/server/rust/oak_runtime/src/node/wasm.rs

```diff
@@ -0,0 +1,673 @@
+const CHANNEL_WRITE: usize = 4;
+const CHANNEL_READ: usize = 5;
+const WAIT_ON_CHANNELS: usize = 6;
+
+/// TODO: doc
+/// TODO: randomized handle allocator
+struct WasmInterface {
+    runtime: Arc<Runtime>,
+
+    readers: HashMap<u64, ChannelReader>,
+    writers: HashMap<u64, ChannelWriter>,
+
+    next_handle: u64,
+
+    memory: Option<wasmi::MemoryRef>,
+}
+
+impl WasmInterface {
+    /// TODO: doc
+    pub fn new(runtime: &Arc<Runtime>, initial_handle: ChannelReader) -> WasmInterface {
+        WasmInterface {
+            runtime: runtime.clone(),
+            readers: {
+                let mut r = HashMap::new();
+                r.insert(1, initial_handle);
+                r
+            },
+            writers: HashMap::new(),
+            next_handle: 2,
+            memory: None,
+        }
+    }
+
+    /// TODO: doc
+    fn node_create(
+        &self,
+        name_ptr: u32,
+        name_length: usize,
+        entrypoint_ptr: u32,
+        entrypoint_length: usize,
+        initial_handle: u64,
+    ) -> Result<(), OakStatus> {
+        let memory = self
+            .memory
+            .as_ref()
+            .expect("WasmInterface memory not attached!?");
+
+        let config_name = memory.get(name_ptr, name_length).map_err(|err| {
+            error!(
+                "node_create: Unable to read name from guest memory: {:?}",
+                err
+            );
+            OakStatus::ERR_INVALID_ARGS
+        })?;
+
+        let config_name = String::from_utf8(config_name).map_err(|err| {
+            error!("node_create: Unable to parse config_name: {:?}", err);
+            OakStatus::ERR_INVALID_ARGS
+        })?;
+
+        let entrypoint = memory
+            .get(entrypoint_ptr, entrypoint_length)
+            .map_err(|err| {
+                error!(
+                    "node_create: Unable to read entrypoint from guest memory: {:?}",
+                    err
+                );
+                OakStatus::ERR_INVALID_ARGS
+            })?;
+
+        let entrypoint = String::from_utf8(entrypoint).map_err(|err| {
+            error!("node_create: Unable to parse entrypoint: {:?}", err);
+            OakStatus::ERR_INVALID_ARGS
+        })?;
+
+        let handle = self.readers.get(&initial_handle).ok_or(()).map_err(|_| {
+            error!("node_create: Invalid handle");
+            OakStatus::ERR_BAD_HANDLE
+        })?;
+
+        Runtime::node_create(&self.runtime, &config_name, &entrypoint, handle.clone()).map_err(
+            |_| {
+                error!("node_create: Config \"{}\" not found", config_name);
+                OakStatus::ERR_INVALID_ARGS
+            },
+        )
+    }
+
+    /// TODO: doc
+    fn channel_create(&mut self, write_addr: u32, read_addr: u32) -> Result<(), OakStatus> {
+        let (writer, reader) = crate::channel::new();
+
+        let write_handle = self.next_handle;
+        self.writers.insert(write_handle, writer);
+        self.next_handle += 1;
```

Comment:
This reverts 10b2015bcb2324d4614393f9504cc

Reply:
Fixed